File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/MessageListUseCase.java
package com.smallchat.backend.chat.application.usecase;

import com.smallchat.backend.chat.application.inputport.MessageListInputPort;
import com.smallchat.backend.chat.application.outputport.MessageOutputPort;
import com.smallchat.backend.chat.domain.model.vo.Message;
import com.smallchat.backend.chat.framework.web.dto.MessageListDto;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

import static com.smallchat.backend.chat.framework.mongodb_adapter.MessageOutputAdapter.MESSAGE_PAGE_LIMIT;

@Service
@RequiredArgsConstructor
public class MessageListUseCase implements MessageListInputPort {
    private final MessageOutputPort messageOutputPort;

    @Override
    public MessageListDto.Response getMessageList(String chatID, Long nextCursor) {
        List<Message> list = messageOutputPort.getMessageList(chatID, nextCursor);
        long newNextCursor = list.isEmpty() || list.size() < MESSAGE_PAGE_LIMIT ? 0L : list.get(0).getCreatedAt().toEpochSecond();
        return new MessageListDto.Response(list.stream().map(Message::toMessageBasicInfo).toList()
                , newNextCursor);
    }
}

File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/MessageListUseCase.kt
package com.smallchat.backend.chat.application.usecase

import com.smallchat.backend.chat.infrastructure.mongodb_adapter.MongoTemplateAdapter
import com.smallchat.backend.chat.interfaces.web.dto.MessageListDto
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service

@Service
@RequiredArgsConstructor
class MessageListUseCase(val mongoTemplateAdapter: MongoTemplateAdapter) {
    fun execute(chatId: String, nextCursor: Long?): MessageListDto.Response {
        val list = mongoTemplateAdapter.getMessageList(chatId, nextCursor)
        val nextCursor = list.firstOrNull()
            ?.sentAt
            ?.epochSecond
        return MessageListDto.Response(list, nextCursor)
    }
}
